### PR TITLE
Test indexer txs scenarios

### PR DESCRIPTION
**Pull request by miiu96**
Created a new test indexer for the semi-integration tests in order to be able to verify if the transaction fee and the gas used field is correctly calculated.
This new indexer is used in the semi-integration tests.

### CODE REVIEW

File: integrationTests/vm/txsFee/multiShard/scCalls_test.go [hunk 1]
```diff
@@ -129,6 +151,15 @@ func TestScCallExecuteOnSourceAndDstShardInvalidOnDst(t *testing.T) {
 	developerFees := testContextSource.TxFeeHandler.GetDeveloperFees()
 	require.Equal(t, big.NewInt(0), developerFees)
 
+	intermediateTxs := testContextSource.GetIntermediateTransactions(t)
```
Comment (bogdan-rosianu) [inline integrationTests/vm/txsFee/multiShard/scCalls_test.go:154]: typo "increment" on line 135?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: integrationTests/mock/databaseWritterMock.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	return nil
+}
+
+// CheckAndCreateAlias -
+func (dwm *DatabaseWriterStub) CheckAndCreateAlias(_ string, _ string) error {
+	return nil
+}
+
+// CheckAndCreateTemplate -
+func (dwm *DatabaseWriterStub) CheckAndCreateTemplate(_ string, _ *bytes.Buffer) error {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/mock/databaseWritterMock.go:62]: done.

File: integrationTests/mock/epochStartNotifierStub.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+type EpochStartNotifierStub struct {
+	RegisterHandlerCalled   func(handler epochStart.ActionHandler)
+	UnregisterHandlerCalled func(handler epochStart.ActionHandler)
+	NotifyAllCalled         func(hdr data.HeaderHandler)
+	NotifyAllPrepareCalled  func(hdr data.HeaderHandler, body data.BodyHandler)
+	epochStartHdls          []epochStart.ActionHandler
+}
+
+// RegisterHandler -
+func (esnm *EpochStartNotifierStub) RegisterHandler(handler epochStart.ActionHandler) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/mock/epochStartNotifierStub.go:18]: done.

File: integrationTests/vm/testIndexer.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	return esProcessor
+}
+
+// SaveTransaction -
+func (ti *testIndexer) SaveTransaction(
+	tx data.TransactionHandler,
+	mbType block.Type,
+	intermediateTxs []data.TransactionHandler,
+) {
+
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/testIndexer.go:124]: done.

File: integrationTests/vm/testIndexer.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		txsPool[string(intTxHash)] = intTx
+	}
+
+	header := &block.Header{
+		ShardID: ti.shardCoordinator.SelfId(),
+	}
+
+	ti.indexer.SaveBlock(blk, header, txsPool, nil, nil, nil)
+
+	// wait --> dataDispatcher needs needs to save the block
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/testIndexer.go:171]: done.

File: integrationTests/vm/txsFee/dns_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	_, _ = vm.CreateAccount(testContext.Accounts, rcvAddr, 0, senderBalance)
+
+	userName := utils.GenerateUserNameForMyDNSContract()
+	txData := []byte("register@" + hex.EncodeToString(userName))
+	// create user name for sender
+	tx := vm.CreateTransaction(0, big.NewInt(0), sndAddr, scAddress, gasPrice, gasLimit, txData)
+	retCode, err := testContext.TxProcessor.ProcessTransaction(tx)
+	require.Equal(t, vmcommon.Ok, retCode)
+	require.Nil(t, err)
+	require.Equal(t, nil, testContext.GetLatestError())
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/dns_test.go:42]: done.

File: integrationTests/vm/txsFee/dns_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	utils.CleanAccumulatedIntermediateTransactions(t, &testContext)
+
+	// create user name for receiver
+	rcvUserName := utils.GenerateUserNameForMyDNSContract()
+	txData = []byte("register@" + hex.EncodeToString(rcvUserName))
+	tx = vm.CreateTransaction(0, big.NewInt(0), rcvAddr, scAddress, gasPrice, gasLimit, txData)
+	retCode, err = testContext.TxProcessor.ProcessTransaction(tx)
+	require.Equal(t, vmcommon.Ok, retCode)
+	require.Nil(t, err)
+	require.Equal(t, nil, testContext.GetLatestError())
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/dns_test.go:69]: done.

File: integrationTests/vm/txsFee/moveBalance_test.go [hunk 8]
```diff
@@ -193,3 +216,39 @@ func TestMoveBalanceMoreGasThanGasLimitPerBlock(t *testing.T) {
...
+	sndAddr := []byte("12345678901234567890123456789012")
+	rcvAddr := []byte("12345678901234567890123456789022")
+	senderNonce := uint64(0)
+	senderBalance := big.NewInt(10000)
+	gasPrice := uint64(10)
+	gasLimit := uint64(100)
+
+	_, _ = vm.CreateAccount(testContext.Accounts, sndAddr, 0, senderBalance)
+	tx := vm.CreateTransaction(senderNonce, big.NewInt(100), sndAddr, rcvAddr, gasPrice, gasLimit, []byte("aaaa"))
+	tx.SndUserName = []byte("invaidUserName")
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/moveBalance_test.go:233]: done.

File: integrationTests/vm/txsFee/multiShard/scCalls_test.go (hunk 1 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/multiShard/scCalls_test.go:154]: is not typo because I want to call a method that does not exist.

**Review by sasurobert — DISMISSED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by mihaib79 — APPROVED** [review]

File: integrationTests/vm/txsFee/relayedESDT_test.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	rtx := vm.CreateTransaction(0, innerTx.Value, relayerAddr, sndAddr, gasPrice, rTxGasLimit, rtxData)
+
+	retCode, err := testContext.TxProcessor.ProcessTransaction(rtx)
+	require.Equal(t, vmcommon.Ok, retCode)
+	require.Nil(t, err)
+
+	_, err = testContext.Accounts.Commit()
+	require.Nil(t, err)
+
+	expectedBalanceSnd := big.NewInt(99999900)
```
Comment (bogdan-rosianu) [inline integrationTests/vm/txsFee/relayedESDT_test.go:43]: replace hardcoded with the difference between `esdtBalance` and the value of the transfer

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: integrationTests/vm/txsFee/asyncCall_test.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+	"math/big"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+)
+
+func TestAsyncCallShoutWork(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/asyncCall_test.go:14]: done

File: integrationTests/vm/txsFee/asyncESDT_test.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+func TestAsyncESDTCallSecondScRefusesPayment(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMs(t, vm.ArgEnableEpoch{})
+	defer testContext.Close()
+
+	egldBalance := big.NewInt(100000000)
+	ownerAddr := []byte("12345678901234567890123456789010")
+	_, _ = vm.CreateAccount(testContext.Accounts, ownerAddr, 0, egldBalance)
+
+	// create a address with ESDT token
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/asyncESDT_test.go:84]: done

File: integrationTests/vm/txsFee/asyncESDT_test.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+func TestAsyncESDTCallsOutOfGas(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMs(t, vm.ArgEnableEpoch{})
+	defer testContext.Close()
+
+	egldBalance := big.NewInt(100000000)
+	ownerAddr := []byte("12345678901234567890123456789010")
+	_, _ = vm.CreateAccount(testContext.Accounts, ownerAddr, 0, egldBalance)
+
+	// create a address with ESDT token
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/asyncESDT_test.go:145]: done

File: integrationTests/vm/txsFee/relayedAsyncESDT_test.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+
+func TestRelayedAsyncESDTCallShouldWork(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMs(t, vm.ArgEnableEpoch{})
+	defer testContext.Close()
+
+	egldBalance := big.NewInt(100000000)
+	ownerAddr := []byte("12345678901234567890123456789010")
+	_, _ = vm.CreateAccount(testContext.Accounts, ownerAddr, 0, egldBalance)
+
+	// create a address with ESDT token
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedAsyncESDT_test.go:23]: done

File: integrationTests/vm/txsFee/relayedAsyncESDT_test.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+func TestRelayedAsyncESDTCall_InvalidCallFirstContract(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMs(t, vm.ArgEnableEpoch{})
+	defer testContext.Close()
+
+	egldBalance := big.NewInt(100000000)
+	ownerAddr := []byte("12345678901234567890123456789010")
+	_, _ = vm.CreateAccount(testContext.Accounts, ownerAddr, 0, egldBalance)
+
+	// create a address with ESDT token
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedAsyncESDT_test.go:89]: done.

File: integrationTests/vm/txsFee/testdata/second/async.export [hunk 15]
```diff
@@ -0,0 +1,4 @@
+init
+doSomething
+callBack
+get
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/testdata/second/async.export:4]: done.

File: integrationTests/vm/txsFee/utils/utilsESDT.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+	esdtData := &esdt.ESDigitalToken{
+		Value:      esdtValue,
+		Properties: []byte{},
+	}
+
+	esdtDataBytes, err := protoMarshalizer.Marshal(esdtData)
+	require.Nil(t, err)
+
+	key := append([]byte("ELRONDesdt"), tokenIdentifier...)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/utils/utilsESDT.go:44]: done.

File: integrationTests/vm/txsFee/asyncCall_test.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
+package txsFee
+
+import (
+	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/asyncCall_test.go:5]: done.

File: integrationTests/vm/txsFee/relayedESDT_test.go (hunk 9 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedESDT_test.go:43]: I prefer the hardcoded values for tests, also everywhere in these tests I use that pattern.

File: integrationTests/vm/txsFee/relayedAsyncESDT_test.go [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+	rtx := vm.CreateTransaction(0, innerTx.Value, relayerAddr, sndAddr, gasPrice, rTxGasLimit, rtxData)
+
+	retCode, err := testContext.TxProcessor.ProcessTransaction(rtx)
+	require.Equal(t, vmcommon.Ok, retCode)
+	require.Nil(t, err)
+
+	_, err = testContext.Accounts.Commit()
+	require.Nil(t, err)
+
+	utils.CheckESDTBalance(t, &testContext, firstSCAddress, token, big.NewInt(2500))
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedAsyncESDT_test.go:62]: I prefer the hardcoded values for tests, also everywhere in these tests I use that pattern.

**Review by sasurobert — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by mihaib79 — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]

**Pull request merged**